Clang 21 complains about implicit casting between char32_t and char16_t throughout the codebase

mcatanzaro opened the pull request:
#### 6a5e969cf2e9ca600e2af4915f692fe3b3492bb7
<pre>
Clang 21 complains about implicit casting between char32_t and char16_t throughout the codebase
<a href="https://bugs.webkit.org/show_bug.cgi?id=299689">https://bugs.webkit.org/show_bug.cgi?id=299689</a>

Reviewed by NOBODY (OOPS!).

This pull request is a WIP draft. It is incomplete and not ready for review.

This prepares WebKit&apos;s cross-platform code for Clang 21&apos;s
-Wcharacter-conversion. In practice, this warning is emitted whenever we
implicitly convert from char32_t to char16_t, or vice-versa. In general,
the solution is to add a cast.

Casting to char16_t to char32_t is safe if you&apos;re certain that none of
the code units are surrogates, so let&apos;s add a cast that asserts this is
true.

Converting from char32_t to char16_t is safe if you&apos;re certain that the
character is in the Basic Multilingual Plane, so let&apos;s add a cast that
asserts this is true.

Since char32_t is not compatible with ICU&apos;s UChar32, and we want to use
char32_t to represent UTF-32, I&apos;ve added variants of various ICU
conversion macros that return char32_t instead of UChar32. They also
receive inputs as spans, use char8_t instead of UChar8, and notably
avoid triggering -Wunsafe-buffer-usage.

In a few cases, I change behavior where the original code looks wrong
and seems easy to fix, e.g. isStrWhiteSpace in ParseInt.h, or
URLParser::isForbiddenHostCodePoint.

EncodingTables.h uses templates throughout the file and I&apos;m not sure how
to fix it, so I decided not to worry about it and just suppressed the
warning. I made the same decision for all of TestWebKitAPI to avoid
requiring changes to GoogleTest.

Lastly, I have added a deprecated/broken cast for use in cases that look
suspicious, where I suspect that the assertion might be hit. These will
require follow-up investigation.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6a5e969cf2e9ca600e2af4915f692fe3b3492bb7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132126 "37 style errors") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4619 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43148 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139641 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84043 "Hash 6a5e969c for PR 51527 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133996 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4567 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4380 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/139641 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/84043 "Hash 6a5e969c for PR 51527 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135072 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/155/builds/4567 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/138/builds/43148 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/139641 "Hash 6a5e969c for PR 51527 does not build (failure)") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/155/builds/4567 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/138/builds/43148 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82860 "Hash 6a5e969c for PR 51527 does not build (failure)") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124189 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/155/builds/4567 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/138/builds/43148 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142287 "Hash 6a5e969c for PR 51527 does not build (failure)") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130633 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4288 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/138/builds/43148 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/142287 "Hash 6a5e969c for PR 51527 does not build (failure)") | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4369 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/156/builds/4380 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/142287 "Hash 6a5e969c for PR 51527 does not build (failure)") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/138/builds/43148 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57534 "Hash 6a5e969c for PR 51527 does not build (failure)") | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4342 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/138/builds/43148 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163600 "Hash 6a5e969c for PR 51527 does not build (failure)") | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4173 "Hash 6a5e969c for PR 51527 does not build (failure)") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67788 "Failed to build and analyze WebKit") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/163600 "Hash 6a5e969c for PR 51527 does not build (failure)") | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4433 "Hash 6a5e969c for PR 51527 does not build (failure)") | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4301 "Hash 6a5e969c for PR 51527 does not build (failure)") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d197488558ef49c020e93203297c7aa6246727dc. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51527

darinadler commented:
One other thought: Not too long after we make these changes we’ll need to make sure we have an EWS that is running clang 21, otherwise we’ll keep re-introducing more of these issues.

mcatanzaro commented:
> One other thought: Not too long after we make these changes we’ll need to make sure we have an EWS that is running clang 21, otherwise we’ll keep re-introducing more of these issues.

Certainly would be nice to have. (Fortunately, Ross is already working on this!) That said, we encounter new warnings almost every time we have a new GCC version and quite often for new Clang versions, and we make do with or without an EWS.

mcatanzaro commented:
EWS noticed that this breaks FontCairoHarfbuzzNG.cpp at least. I'll follow up on any other failures tomorrow.

darinadler commented:
Oops, I did "comment" but I meant "request changes".

rkirsling commented:
Please feel free to revert #51463 as part of this.

mcatanzaro commented:
I'll move this to Draft status, since I need time to respond to Darin's feedback and review the test failures.

mcatanzaro commented on Source/JavaScriptCore/yarr/YarrPattern.cpp:
```diff
@@ -287,8 +287,8 @@ class CharacterClassConstructor {
                 char16_t ch;
-                for (auto* set = canonicalCharacterSetInfo(info->value, m_canonicalMode); (ch = *set); ++set)
-                    addSorted(ch);
+                for (auto* set = canonicalCharacterSetInfo(info->value, m_canonicalMode); (ch = static_cast<char16_t>(*set)); ++set)
+                    addSorted(static_cast<char32_t>(ch));
```
I'm pretty sure this doesn't introduce any bug. It's preserving the original behavior; the casts just make it more obvious.

But I don't doubt that it's wrong. I added the casts without thinking too much about it simply because the original code explicitly used char16_t here.

mcatanzaro commented:
I'm going to explore how hard it is to add the safer conversion functions suggested by Darin.

I'm still thinking about whether we can safely use `ASSERT()` though. We should use `ASSERT()` as if it were enabled in release builds. (Regardless of whether it is actually enabled today or not, we ought to attempt to enable asserts in the future.) Processing invalidly-encoded text is not necessarily a WebKit bug; WebKit shouldn't _normally_ encounter invalidly-encoded text, but also shouldn't crash on it either. However, we might be able to add asserts anyway by declaring that code that processes untrusted input should deal with ensuring the input is validly-encoded before passing it on to WebKit's string-handling functions.

darinadler commented:
My suggestion is to have unsafe versions of the functions which do not assert, basically a FIXME. And versions that are used when the code has already checked for surrogates or characters outside the BMP range, useful in many many cases, and they’d have assertions. And versions that intentionally handle surrogates and non-BMP characters in some predictable way, like returning `nullopt` or `U+FFFD` or even leaving the surrogate as a `char32_t` if that’s what we find useful, so maybe identical to the unsafe version, just with a different name.

In my suggestion we would not use a `static_cast` directly, not just because it’s ambiguous which of the above situations it is, but also because a `static_cast` can do conversions involving other types, like convert a `bool` to a `char32_t`, and it’s nice to know that’s not what’s happening, it’s just `char16_t` and `char32_t` (and whatever other types makes sense for the semantics of the functions we invent).

darinadler commented:
But if you want to land this with `static_cast`, I am happy to do another pass afterward and introduce those functions.

mcatanzaro commented on Source/JavaScriptCore/yarr/YarrSyntaxChecker.cpp:
```diff
@@ -32,6 +32,8 @@
 namespace JSC { namespace Yarr {
 
+// This class must match the YarrParser Delegate interface documented in
+// Yarr::parse in YarrParser.h.
```
The comment is not actually needed anymore since [290702@main](https://commits.webkit.org/290702@main). There is a static assert to ensure the class is correct.

[293609@main](https://commits.webkit.org/293609@main) unnecessarily reintroduced a comment describing the interface, which diverged from the actual implementation. We should simply re-remove the comment since the concept is clearly better. I will follow up on that.

mcatanzaro commented:
I'm more confident in the following "casts" that I added in a new file, wtf/unicode/CharacterCasts.h:

```
// This only works if you're certain the input is in the Basic Multilingual Plane. If you need to
// handle arbitrary values, use U16_LEAD() and U16_TRAIL() instead.
inline constexpr char16_t castBMPUTF32CodeUnitToUTF16(char32_t c)
{
    ASSERT_UNDER_CONSTEXPR_CONTEXT(U_IS_BMP(c));
    return static_cast<char16_t>(c);
}

// FIXME: Code using this is probably broken. Remove it.
inline constexpr char16_t deprecatedBrokenCastUTF32CodeUnitToUTF16IgnoringSurrogates(char32_t c)
{
    return static_cast<char16_t>(c);
}

// This only works if you're certain the input is not a surrogate. Be careful, because you're
// responsible for handling even *invalid* surrogates before calling this function, including
// unpaired surrogates or surrogate trail before the lead.
//
// To handle arbitrary values, which may be surrogates, use U16_GET_SUPPLEMENTARY(), or use
// u16Get from UnicodeExtras.h, or use SurrogatePairAwareTextIterator.
inline constexpr char32_t castNonSurrogateUTF16CodeUnitToUTF32(char16_t c)
{
    ASSERT_UNDER_CONSTEXPR_CONTEXT(!U16_IS_SURROGATE(c));
    return static_cast<char32_t>(c);
}

// FIXME: Code using this is probably broken. Remove it.
inline constexpr char32_t deprecatedBrokenCastUTF16CodeUnitToUTF32IgnoringSurrogates(char16_t c)
{
    return static_cast<char32_t>(c);
}

// This only works if the input is ASCII (not Latin-1).
template <typename CharType>
constexpr char32_t castSingleByteCodeUnitToUTF32(CharType c)
{
    static_assert(sizeof(CharType) == 1);
    static_assert(!std::is_same_v<CharType, Latin1Character>);

    ASSERT_UNDER_CONSTEXPR_CONTEXT(U8_IS_SINGLE(c));
    return static_cast<char32_t>(c);
}

// FIXME: Code using this is probably broken for non-ASCII Latin-1 characters (0x80..0xff). Remove it.
constexpr char32_t deprecatedBrokenCastLatin1CharacterToUTF32WithoutConvertingToUnicode(Latin1Character c)
{
    return static_cast<char32_t>(c);
}
```

I'm going to push my WIP, unfinished, not ready for review changes now, just for my own benefit, so I do not completely lose them. Then unless you suggest otherwise, my plan for next week is to scrap all my ICU wrapper functions and start over with a much more minimal commit, introducing these new casts.

> But finally I noticed a problem: these functions return negative values to indicate error conditions. They can also return U_SENTINEL which is -1. char32_t is an unsigned type, whereas UChar32 is a signed type. So my work is broken. It really does have to be UChar32, unfortunately.

Actually, I'm afraid we _already_ introduced this class of bugs when we replaced UChar32 with char32_t.

mcatanzaro committed:
Clang 21 complains about implicit casting between char32_t and char16_t throughout the codebase
https://bugs.webkit.org/show_bug.cgi?id=299689

Reviewed by NOBODY (OOPS!).

This pull request is a WIP draft. It is incomplete and not ready for review.

This prepares WebKit's cross-platform code for Clang 21's
-Wcharacter-conversion. In practice, this warning is emitted whenever we
implicitly convert from char32_t to char16_t, or vice-versa. In general,
the solution is to add a cast.

Casting to char16_t to char32_t is safe if you're certain that none of
the code units are surrogates, so let's add a cast that asserts this is
true.

Converting from char32_t to char16_t is safe if you're certain that the
character is in the Basic Multilingual Plane, so let's add a cast that
asserts this is true.

Since char32_t is not compatible with ICU's UChar32, and we want to use
char32_t to represent UTF-32, I've added variants of various ICU
conversion macros that return char32_t instead of UChar32. They also
receive inputs as spans, use char8_t instead of UChar8, and notably
avoid triggering -Wunsafe-buffer-usage.

In a few cases, I change behavior where the original code looks wrong
and seems easy to fix, e.g. isStrWhiteSpace in ParseInt.h, or
URLParser::isForbiddenHostCodePoint.

EncodingTables.h uses templates throughout the file and I'm not sure how
to fix it, so I decided not to worry about it and just suppressed the
warning. I made the same decision for all of TestWebKitAPI to avoid
requiring changes to GoogleTest.

Lastly, I have added a deprecated/broken cast for use in cases that look
suspicious, where I suspect that the assertion might be hit. These will
require follow-up investigation.
